include/openthread/thread_ftd.h

```diff
@@ -195,6 +195,59 @@ otError otThreadSetRouterEligible(otInstance *aInstance, bool aEligible);
  */
 otError otThreadSetPreferredRouterId(otInstance *aInstance, uint8_t aRouterId);
 
+/**
+ * This enumeration represents the power supply configuration on a device.
+ *
+ * This is used as a property in `otDeviceConfig` to calculate the leader weight.
+ *
+ */
+typedef enum
+{
+    OT_POWER_SUPPLY_BATTERY           = 0, ///< Battery powered.
+    OT_POWER_SUPPLY_EXTERNAL          = 1, ///< Externally powered (mains-powered).
+    OT_POWER_SUPPLY_EXTERNAL_STABLE   = 2, ///< Stable external power with a battery backup or UPS.
+    OT_POWER_SUPPLY_EXTERNAL_UNSTABLE = 3, ///< Potentially unstable ext power (e.g. light bulb powered via a switch).
+} otPowerSupplyConfig;
+
+/**
+ * This structure represents the device configurations which are used for calculating the local leader weight on a
+ * device.
+ *
+ * The parameters are set based on device's capability, whether acting as border router, its power supply config, etc.
+ *
+ * `mIsUnstable` indicates operational stability of device and is determined via a vendor specific mechanism. It can
+ * include the following cases:
+ *  - Device internally detects that it loses external power supply more often than usual. What is usual is
+ *    determined by the vendor.
+ *  - Device internally detects that it reboots more often than usual. What is usual is determined by the vendor.
+ *
+ */
+typedef struct otDeviceConfig
+{
+    otPowerSupplyConfig mPowerSupply;            ///< Power supply config.
+    bool                mIsBorderRouter : 1;     ///< Whether device is a border router.
+    bool                mSupportsCcm : 1;        ///< Whether device supports CCM (can act as a CCM border router).
+    bool                mIsUnstable : 1;         ///< Operational stability of device (vendor specific).
+    int8_t              mLeaderWeightAdjustment; ///< Weight adjustment. Should be -16 to +16 (clamped otherwise).
+} otDeviceConfig;
```

Comment:
nit on naming - how about something like `otDeviceProperties`? This makes more clear that we are simply informing OT about the device, rather than configuring OT itself.

Reply:
Good idea. `otDeviceProperties` sounds better. Will rename it.

Reply:
Renamed in new push.